File: src/ast/node/block.ts
import { SymbolTable } from "../../lib/symbolTable";
import type { Pass } from "../../pass/pass";
import type { Dependent } from "./expression";
import { Lambda } from "./lambda";
import { Node } from "./node";
import { ExecutionResult, Statement, type Standalone } from "./statement";
import type { Value } from "./value";

export class Block extends Node<Node<any>> {
    public symbolTable = new SymbolTable();
    public returnValue: any | null = null;

    constructor(public statements: Standalone[], parent: Node<any> | null) {
        super(parent);
    }

    public execute(): ExecutionResult {
        for (const statement of this.statements) {
            const result = statement.execute();

            if (statement instanceof Statement && result !== ExecutionResult.ContinueEvaluation)
                return result as ExecutionResult;
        }

        return ExecutionResult.ContinueEvaluation;
    }

    public visit(pass: Pass): void {
        pass.visitBlock(this);
    }

    public getSymbol(name: string): any | null {
        if (this.symbolTable.has(name))
            return this.symbolTable.get(name) ?? null;
        else if (this.parent instanceof Block)
            return this.parent.getSymbol(name);
        else return this.parent?.getParent<Block>(Block)?.getSymbol(name) ?? null;
    }

    public setSymbol(name: string, value: any) {
        if (!this.setSymbolInternal(name, value))
            this.symbolTable.set(name, value);
    }

    private setSymbolInternal(name: string, value: any): boolean {
        if (this.symbolTable.has(name)) {
            this.symbolTable.set(name, value);
            return true;
        } else if (this.parent instanceof Block)
            return this.parent.setSymbolInternal(name, value)
        else return this.parent?.getParent<Block>(Block)?.setSymbolInternal(name, value) ?? false;
    }
}

File: src/main/kotlin/me/honkling/miniscript/parser/ast/Block.kt
package me.honkling.miniscript.parser.ast

import me.honkling.miniscript.MiniScript
import me.honkling.miniscript.diagnostic.MiniScriptException
import me.honkling.miniscript.parser.ast.expression.Expression
import me.honkling.miniscript.parser.ast.stack.Frame
import me.honkling.miniscript.parser.ast.statement.ExecutionResult
import me.honkling.miniscript.parser.ast.statement.Statement
import me.honkling.miniscript.pass.Pass

open class Block(
    val miniScript: MiniScript,
    val statements: MutableList<Node<*>>,
    parent: Node<*>?,
) : Node<Node<*>?>(parent) {
    var returnValue: Any? = null

    fun execute(pushStack: Boolean = true): Pair<Any?, ExecutionResult> {
        if (pushStack) pushToStack()
        returnValue = null

        for ((index, statement) in statements.withIndex()) {
            val isLast = index + 1 == statements.size

            when (statement) {
                is Expression<*> -> {
                    val result = statement.get()

                    if (isLast)
                        return result
                }
                is Statement -> {
                    val result = statement.execute()

                    if (result != ExecutionResult.ContinueExecution) {
                        if (pushStack) popFromStack()
                        return null to result
                    }
                }
                else -> throw MiniScriptException.RuntimeError("Expected a statement or expression in block")
            }
        }

        if (pushStack) popFromStack()
        return null to ExecutionResult.ContinueExecution
    }

    fun pushToStack(): Frame {
        val frame = Frame()
        miniScript.environment.executionStack += frame
        return frame
    }

    fun popFromStack() {
        miniScript.environment.executionStack.removeLast()
    }

    override fun accept(pass: Pass) {
        pass.visitBlock(this)
    }
}

class SymbolTable : HashMap<String, Any?>()

interface SymbolHolder {
    val symbolTable: SymbolTable
}